### PR TITLE
ssh/tailssh: allow multiple sessions on the same conn

### LINKED ISSUE #4920: Cannot open multiple ssh sessions to tailscale nodes with --ssh=on and ControlMaster=on
### What is the issue?

With `ControlMaster` configured on my local Mac, opening up another session to a tailscale node (with `--ssh` enabled) will fail and break the existing connected session.

Removing `ControlMaster on` configuration from my local `~/.ssh/config` fixes the issue.

### Steps to reproduce

Ensure ControlMaster is enabled:
```bash
cat ~/.ssh/config

ControlMaster auto
ControlPath ~/.ssh/cache/%C
ControlPersist 6h
```

Open a new ssh connection to tailscale node (with `--ssh` enabled):

```bash
$ ssh node1
```

In another terminal window, open another ssh session to the same node. The underlying shared connection will terminate:
```bash
$ ssh node1
Shared connection to <node1-tailscale-ip> closed.
```

### Are there any recent changes that introduced the issue?

_No response_

### OS

macOS

### OS version

MacOS Monterey 12.4

### Tailscale version

1.26.1

### Bug report

_No response_

### PR DESCRIPTION
**Pull request by maisem**
Fixes #4920
Fixes tailscale/corp#5633
Updates #4479

Signed-off-by: Maisem Ali <maisem@tailscale.com>

### CODE REVIEW

File: ssh/tailssh/tailssh.go [hunk 1]
```diff
@@ -516,32 +529,47 @@ func (srv *server) fetchPublicKeysURL(url string) ([]string, error) {
...
-		metricPublicKeyAccepts.Add(1)
+		return nil, false
 	}
+	return cr, true
+}
 
+// handlePostSSHAuth runs an SSH session after the SSH-level authentication,
+// but not necessarily before all the Tailscale-level extra verification has
+// completed. It also handles SFTP requests.
+func (c *conn) handlePostSSHAuth(s ssh.Session) {
```
Comment (bradfitz) [inline ssh/tailssh/tailssh.go:562]: handleSessionPostSSHAuth ?

File: ssh/tailssh/tailssh.go [hunk 2]
```diff
@@ -92,13 +91,21 @@ func init() {
...
 	}
-	ss.HandleConn(c)
+	srv.mu.Lock()
+	mak.Set(&srv.activeConns, c, true)
+	srv.mu.Unlock()
+	defer func() {
+		srv.mu.Lock()
+		delete(srv.activeConns, c)
+		srv.mu.Unlock()
+	}()
```
Comment (bradfitz) [inline ssh/tailssh/tailssh.go:107]: move to method on on server, so this is like:

```go
    s.trackActiveConn(c, true) // add
    defer c.trackActiveConn(c, false) // remove
```

?

🤷‍♂️

**Review by bradfitz — COMMENTED** [review]
Docs on all the things that "session ID" means too somewhere?

File: ssh/tailssh/tailssh.go [hunk 3]
```diff
@@ -135,25 +150,33 @@ func (srv *server) OnPolicyChange() {
...
 	info         *sshConnInfo // set by setInfo
 	localUser    *user.User   // set by checkAuth
 	userGroupIDs []string     // set by checkAuth
 
-	insecureSkipTailscaleAuth bool // used by tests.
+	mu sync.Mutex // protects the following
+	// idH is the RFC4253 sec8 hash H. It is used to identify the connection,
+	// and is shared among all sessions. It should not be shared outside
+	// process.
+	// It is reffered to as SessionID by the gliderlabs/ssh library.
```
Comment (bradfitz) [inline ssh/tailssh/tailssh.go:170]: referred

Comment (bradfitz) [reply ssh/tailssh/tailssh.go:170]: It is confusingly referred to as ...

File: ssh/tailssh/tailssh.go [hunk 4]
```diff
@@ -726,16 +775,18 @@ func (c *conn) isStillValid(pubKey ssh.PublicKey) bool {
 
 // checkStillValid checks that the session is still valid per the latest SSHPolicy.
 // If not, it terminates the session.
```
Comment (bradfitz) [inline ssh/tailssh/tailssh.go:777]: any/all active sessions

**Review by bradfitz — APPROVED** [review]

**Comment by sailorfrag** [comment]
Yay thanks

**Pull request merged**